enhancements/cloud-integration/tokenized-auth-enablement-operators-on-cloud.md

```diff
@@ -0,0 +1,503 @@
+under the Red Hat brand (that need to reach cloud APIs).
+ 
+As the managed services, including HyperShift teams, offering a downstream opinionated, supported and managed lifecycle 
+of OpenShift (in the forms of ROSA, ARO, OSD on GCP, HyperShift, etc), the OpenShift platform should have as close as 
+possible, native integration with core platform operators when clusters use tokenized cloud auth, driving the use of 
+layered products.
+
+As the Hypershift team, where the only credential mode for clusters/customers is STS (on AWS) , the Red Hat branded 
+Operators that must reach the AWS API, should be enabled to work with STS credentials in a consistent, and automated 
+fashion that allows customer to use those operators as easily as possible, driving the use of layered products.
+
+As an operator team that may already be supporting cloud credential provisioning, including a manner for allowing
+operand instances to have a unique set of credential per instance, I want my solution to keep working (Quay Operator).
+
+As a customer of OpenShift layered products I can confidently upgrade the operator even if the next version requires
+elevated permissions be provided.
+
+As an operator author I have a set of best practices and tools that help me develop a compliant operator.
+
+As a monitor of OpenShift layered products I can identify which operators are compliant with STS best practices and 
+tools. 
+
+As a cluster admin, it's clear when installing an operator if it is STS compliant (i.e. either doesn't need cloud 
+resources or does and is following the best practices).
+
+As an OpenShift user I want the STS enablement solution to be as cloud-agnostic as possible.
+
+### Goals
+
+Allow Day-2 Operators to access cloud provider resources as seamlessly as possible when being installed, used and
+updated on STS enabled clusters.
+
+While providing the above goal, allow for multi-tenancy. In this sense multi-tenancy means that an operator may
+provide different cloud provider credentials per operand such that operator users may uniquely access non-shared cloud
+resources.
+
+Operator authors have a way to notify, guide, and assist Day-2 Operator admins in providing the required cloud provider credentials matched to their permission
+needs for install and update.
+
+Ideally, a solution here will work in both HyperShift (STS always) and non-HyperShift clusters.
+
+### Non-Goals
+
+Day 1 operators are not included in this proposal, they are pre-provisioned for STS enabled clusters (at least 
+for HyperShift)
+
+
+
+## Proposal
+
+**Cloud Credential Operator (CCO) changes**: Add a "Token" mode, distinct from disabled/manual mode that will look for 
+and process CredentialRequests referenced in Operator CRs. This will operate in one of two ways depending on which other
+parts of this EP for other components are adopted: specifically HyperShift/Rosa adding CCO with Token mode or continuing
+with the current support for the pod identity webhook.
```

Comment:
I think we should avoid adding a `"Token"` `credentialsMode` unless our intent is for existing AWS STS, GCP Workload Identity Federation and (soon) Azure Workload Identity clusters to start utilizing this mode in the installer & documentation because I worry it would be confusing to users otherwise. Also, the credentials are manually maintained by the user so `"Manual"` makes sense as a `credentialsMode`.

Would it be reasonable (and possible?) for operators & the console to determine that a cluster is STS-enabled by checking the cluster `authentication`, `cloudcredential` and `infrastructure` config CRs?

For example,

```
# infrastructure platform is AWS
$ oc get infrastructure cluster -o jsonpath --template '{ .status.platform }'
AWS

# credentialsMode is "Manual"
$ oc get cloudcredential cluster -o jsonpath --template '{ .spec.credentialsMode }'
Manual

# serviceAccountIssuer is non empty
$ oc get authentication cluster -o jsonpath --template='{ .spec.serviceAccountIssuer }'
abutcher-oidc.s3.us-east-1.amazonaws.com
```

If it will not be possible for an operator to check the above I wonder if it would make sense to add something to the `infrastructure` CR indicating that the cluster is an observed STS mode? I'm not sure if the `infrastructure` CR works like that.

Reply:
Not adding a "Token" mode has much bigger implications than how we determine if the a cluster is in STS mode..it means someone/something else becomes responsible for creating the secret containing the ARN+tokenpath.  It also means we lack an easy way to validate+report whether the ARN in question has the requested permissions.

Reply:
When CCO starts up it disables the CredentialsRequests controller when the `credentialsMode` is `"Manual"`. I think that we could enable the controller in all `credentialsMode`s but only reconcile CRs in `"Manual"` mode when they have the spec field (roleARN, GCP/Azure equivalents) for brokering secret creation.

Reply:
I think we just assume that if CredentialRequest `.spec.roleARN` or `.spec.GCPRole` or `.spec.AzureRole` have been set we're in token mode. If none of those, or more than one are set, we do nothing emitting an error in the latter.

In one sense the admin manually created the role and told us about it, but CCO is definitely actively engaged in creating the config secret. Overall I'm not too worried about manual mode no longer being inert but looking for spec values unique per cloud and keying off of that.

Reply:
yep, adopted yours and Andrew's suggestions around this. Quoted Andrew's triple check test and started referring to this CCO being token-aware everywhere in the doc.

Reply:
Can we clarify somewhere what the differences in behaviour are between manual mode pre-this enhancement and post, in particular it would be good to understand what happens when: